### PR TITLE
fixed rosetta issues

### PR DESCRIPTION
**Pull request by jhd2best**

### CODE REVIEW

File: hmy/blockchain.go [hunk 1]
```diff
@@ -161,18 +161,19 @@ func (hmy *Harmony) GetPreStakingBlockRewards(
...
-			return nil, fmt.Errorf("could not find receipt for tx: %v", stx.Hash().String())
-		}
-		if len(receipts) <= int(receiptIndex) {
-			return nil, fmt.Errorf("invalid receipt indext %v (>= num receipts: %v) for tx: %v",
-				receiptIndex, len(receipts), stx.Hash().String())
-		}
-		txFee := new(big.Int).Mul(stx.GasPrice(), big.NewInt(int64(receipts[receiptIndex].GasUsed)))
-		txFees = new(big.Int).Add(txFee, txFees)
-	}
+	// todo current not add StakingTransactions tx fee to leader
```
Comment (LeoHChen) [inline hmy/blockchain.go:164]: any impact removing this part of code? why?

File: .github/workflows/ci-rosetta.yaml [hunk 2]
```diff
@@ -0,0 +1,88 @@
+name: release harmony
+
+on:
+  push:
+    branches: ['rosetta_test']
```
Comment (LeoHChen) [inline .github/workflows/ci-rosetta.yaml:5]: is it the right branch?

Comment (jhd2best) [reply .github/workflows/ci-rosetta.yaml:5]: this ci is only for test. will remove after all things done

File: hmy/blockchain.go (hunk 1 above)
Comment (jhd2best) [reply hmy/blockchain.go:164]: already checked with RJ. the gas fee of staking transaction never reward to leader. will delete this segment then

File: core/vm/interpreter.go [hunk 3]
```diff
@@ -264,14 +264,24 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 			mem.Resize(memorySize)
 		}
 
+		var afterHook HookAfter
```
Comment (LeoHChen) [inline core/vm/interpreter.go:267]: this file needs review by @rlan35

File: hmy/tracer.go [hunk 4]
```diff
@@ -813,6 +815,40 @@ func (hmy *Harmony) ComputeTxEnv(block *types.Block, txIndex int, reexec uint64)
 	return nil, vm.Context{}, nil, fmt.Errorf("transaction index %d out of range for block %#x", txIndex, block.Hash())
 }
 
+// ComputeTxEnvEachBlockWithoutApply returns the execution environment of a certain transaction.
+func (hmy *Harmony) ComputeTxEnvEachBlockWithoutApply(block *types.Block, reexec uint64, cb func(int, *types.Transaction, core.Message, vm.Context, *state.DB) bool) error {
+	// Create the parent state database
+	parent := hmy.BlockChain.GetBlock(block.ParentHash(), block.NumberU64()-1)
```
Comment (LeoHChen) [inline hmy/tracer.go:821]: how about genesis block?

File: hmy/tracer.go [hunk 5]
```diff
@@ -836,12 +872,82 @@ type StructLogRes struct {
...
+	if r.Memory != nil {
+		return r.Memory
+	}
+
+	if r.rawMemory != nil {
+		memory := make([]string, 0, (len(r.rawMemory)+31)/32)
+		for i := 0; i+32 <= len(r.rawMemory); i += 32 {
+			memory = append(memory, fmt.Sprintf("%x", r.rawMemory[i:i+32]))
+		}
+		r.Memory = memory
```
Comment (LeoHChen) [inline hmy/tracer.go:927]: will r.Memory == nil?

File: hmy/tracer.go [hunk 6]
```diff
@@ -836,12 +872,82 @@ type StructLogRes struct {
...
+	if r.AfterStack != nil {
+		return r.AfterStack
+	}
+
+	if r.rawAfterStack != nil {
+		stack := make([]string, len(r.rawAfterStack))
+		for i, stackValue := range r.rawAfterStack {
+			stack[i] = hex.EncodeToString(math.PaddedBigBytes(stackValue, 32))
+		}
+		r.AfterStack = stack
```
Comment (LeoHChen) [inline hmy/tracer.go:911]: if r.AfterStack == nil?

File: hmy/tracer.go [hunk 7]
```diff
@@ -836,12 +872,82 @@ type StructLogRes struct {
...
+	if r.Storage != nil {
+		return r.Storage
+	}
+
+	if r.rawStorage != nil {
+		storage := make(map[string]string)
+		for i, storageValue := range r.rawStorage {
+			storage[hex.EncodeToString(i.Bytes())] = hex.EncodeToString(storageValue.Bytes())
+		}
+		r.Storage = storage
```
Comment (LeoHChen) [inline hmy/tracer.go:943]: same thing, what if r.Storage == nil?

**Review by LeoHChen — COMMENTED** [review]
Please add one or a few GitHub issues to describe the problems and reference them in this PR.  That can provide more context to the team.  Please also get review comments from @rlan35 and XP.

File: core/vm/instructions.go [hunk 8]
```diff
@@ -889,6 +889,7 @@ func opStop(pc *uint64, interpreter *EVMInterpreter, contract *Contract, memory
 func opSuicide(pc *uint64, interpreter *EVMInterpreter, contract *Contract, memory *Memory, stack *Stack) ([]byte, error) {
 	balance := interpreter.evm.StateDB.GetBalance(contract.Address())
 	interpreter.evm.StateDB.AddBalance(common.BigToAddress(stack.pop()), balance)
+	stack.push(balance)
```
Comment (peekpi) [inline core/vm/instructions.go:892]: DO NOT change EVM behavior, it's a very dangerous operation.

File: core/vm/interpreter.go [hunk 9]
```diff
@@ -264,14 +264,24 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
...
 		// execute the operation
 		res, err = operation.execute(&pc, in, contract, mem, stack)
 
+		// record the after executed operation content
+		if afterHook != nil {
+			afterHook(mem, stack)
+		}
+
+		if op == SELFDESTRUCT {
+			stack.pop()
```
Comment (peekpi) [inline core/vm/interpreter.go:282]: DO NOT change EVM behavior, even though it's safe here.

File: hmy/tracers/tracer.go [hunk 10]
```diff
@@ -543,7 +543,7 @@ func (jst *Tracer) CaptureStart(env *vm.EVM, from common.Address, to common.Addr
 }
 
 // CaptureState implements the Tracer interface to trace a single step of VM execution.
-func (jst *Tracer) CaptureState(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, memory *vm.Memory, stack *vm.Stack, contract *vm.Contract, depth int, err error) error {
+func (jst *Tracer) CaptureState(env *vm.EVM, pc uint64, op vm.OpCode, gas, cost uint64, memory *vm.Memory, stack *vm.Stack, contract *vm.Contract, depth int, err error) (vm.HookAfter, error) {
```
Comment (peekpi) [inline hmy/tracers/tracer.go:546]: Can afterHook be removed? I can't find where it is used.

File: rosetta/services/tx_operation.go [hunk 11]
```diff
@@ -382,8 +455,9 @@ var (
 	// where the sender is a contract address. This is also known as ops for an 'internal' transaction.
 	// All operations have at least 7 elements on the stack when executed.
 	internalNativeTransferEvmOps = map[string]interface{}{
-		vm.CALL.String():     struct{}{},
-		vm.CALLCODE.String(): struct{}{},
+		vm.CALL.String():         struct{}{},
```
Comment (peekpi) [inline rosetta/services/tx_operation.go:458]: 'Create' and 'Create2' opcode are missing.
'Callcode' does not really transfer Ether.

**Review by peekpi — CHANGES_REQUESTED** [review]

**Review by LeoHChen — APPROVED** [review]

**Pull request merged**